multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java

```diff
@@ -138,4 +154,55 @@ private FileMetadata parseFileEntry(FileEntry fileEntry) {
                                     .namespace(fileEntry.getNamespace())
                                     .build();
     }
+
+    private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String mtarEndpoint) throws FileStorageException, IOException {
+        if (!UriUtil.isValidUrl(mtarEndpoint)) {
+            throw new SLException("Invalid URL");
+        }
+        String decodedUrl = URLDecoder.decode(mtarEndpoint, StandardCharsets.UTF_8);
+        WebClient client = new RestUtil().createWebClient(false);
+
+        ClientResponse response = client.get()
+                                        .uri(decodedUrl)
+                                        .exchange()
+                                        .block();
+        long fileSize = response.headers()
+                                .contentLength()
+                                .orElseThrow(() -> new SLException("MTAR endpoint did not return Content-Length header"));
+        long maxUploadSize = new Configuration().getMaxUploadSize();
+        if (fileSize > maxUploadSize) {
+            throw new SLException(MessageFormat.format(Messages.MAX_UPLOAD_SIZE_EXCEEDED, maxUploadSize));
+        }
+
+        String fileName = extractFileName(decodedUrl);
+        try (InputStream content = response.body(this::extractContent)) {
+            return fileService.addFile(spaceGuid, namespace, fileName, content, fileSize);
+        }
+    }
+
+    private String extractFileName(String url) {
+        int fragmentIdx = url.indexOf('#');
+        if (fragmentIdx != -1) {
+            url = url.substring(0, fragmentIdx);
+        }
+        int queryIdx = url.indexOf('?');
+        if (queryIdx != -1) {
+            url = url.substring(0, queryIdx);
+        }
+        int startIdx = url.lastIndexOf('/') + 1;
+        String fileName = url.substring(startIdx);
+        if (!FileUtils.isFilenameValid(fileName)) {
+            throw new SLException("Invalid MTAR name");
+        }
+        return fileName;
+    }
+
+    private InputStream extractContent(ReactiveHttpInputMessage httpResponse, BodyExtractor.Context context) {
+        return httpResponse.getBody()
+                           .map(dataBuffer -> dataBuffer.asInputStream(true))
+                           .collectList()
+                           .map(list -> new SequenceInputStream(Collections.enumeration(list)))
```

Comment:
I think this will create a SequenceInputStream from multiple ByteArrayInputStreams, which means that the entire file content will be stored in-memory. Can you check, please? If I'm right, find another way to download the file.